### PR TITLE
Adding BiCGSTAB solver for sparse systems

### LINKED ISSUE #2387: Adding sparse solvers for linear systems
**Describe the feature and the current behavior/state.**
Tensorflow doesn't have a direct way to solve sparse linear systems of equations. Specifically: solving Ax=b for x where A is a sparse matrix. In addition, we should be able to propagate gradients, similarly to tf.linalg.solve.

I would like to add a sparse solver (BiCGSTAB for now, we can consider others as well). Initially I was planning on adding this to Tensorflow, but after some internal discussion Tensorflow Addons was suggested instead.

I wanted to give the maintainers a heads-up, and check if there are any concerns/suggestions before I prepare this and send it out for review. Thanks!

**Relevant information**
- Are you willing to contribute it (yes/no): Yes, I have most of the code and need to move it here.
- Are you willing to maintain it going forward? (yes/no): Yes
- Is there a relevant academic paper? (if so, where): BiCGSTAB is a well-established algorithm
- Is there already an implementation in another framework? (if so, where): Yes, e.g. scipy (but no gradients)
- Was it part of tf.contrib? (if so, where): No

**Which API type would this fall under (layer, metric, optimizer, etc.)
I propose adding a new type 'sparse' under which sparse operations outside the scope of core Tensorflow could go. The BiCGSTAB solver could be accessed using e.g., tensorflow_addons.sparse.bicgstab_solver.

**Who will benefit with this feature?**
End users (including myself) who want to solve sparse linear systems, especially within the context of optimization.

There are multiple questions like this:
https://stackoverflow.com/questions/60348419/is-there-an-efficient-way-of-solving-sparse-linear-equations-in-tensorflow-that

**Any other info.**
For now I plan on adding the CPU kernel only. In the future GPU kernel and other solvers could be added if there is demand.

### PR DESCRIPTION
**Pull request by tabakg**
Signed-off-by: Gil Tabak <tabakg@google.com>

# Description

This pull request will add a custom CPU op for solving sparse linear systems using the BiCGSTAB method. This is based on using the Eigen library.

Fixes #2387 

## Type of change

I'd like to start a new subpackage specifically for sparse linear algebra operations not supported (and without plans of being supported) by core Tensorflow. This includes implementation of sparse solvers using iterative methods as custom ops.

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

I used pytest locally, and I'm not familiar enough with the CI workflows here yet to make sure my tests are run - I'm going to open the pull request and integrate tests after I can see how it works.

### CODE REVIEW

**Comment by tabakg** [comment]
Hi maintainers! I was able to finally get the _sparse_ops.so library to build locally, and it came down to carefully going through the steps to install and build (and honestly I'm not entirely sure what the correct sequence of steps is, but it magically appeared at one point). Now I'm having the same issue with the file missing here across the tests. Could you point me in the right direction to what might be missing or needs to be tweaked to get this working? Thank you!

**Comment by WindQAQ** [comment]
Looks like only windows and macos fail. @bhack Can you help this one? I'm busy with other things 😭

**Comment by bhack** [comment]
> Looks like only windows and macos fail. @bhack Can you help this one? I'm busy with other things sob

I suppose it was still on `ecosystem review` and not `feature-approved-for-pr`. There was a similar issue in TF core.

**Comment by tabakg** [comment]
> > Looks like only windows and macos fail. @bhack Can you help this one? I'm busy with other things sob
> 
> I suppose it was still on `ecosystem review` and not `feature-approved-for-pr`. There was a similar issue in TF core.

What is the process for ecosystem review? In an internal discussion it was suggested to me to proceed adding this to TF Addons. Sorry, I'm not familiar with this process.

**Comment by bhack** [comment]
Ecosystem review label it is mainly this https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#requirements-for-new-contributions-to-the-repository
@tomerk Will periodically review `ecosystem reviews` labels with TF core, TF models, Keras and the other ecosystem projects/SIGs about potential conflicts/duplicates with the internals, often "not-public", roadmaps.
We also asked to @tomerk to evaluate if an ecosystem resource could review the PR in another TF ecosystem repository or if there is a resource that could help to review the feature here  e.g. like https://github.com/tensorflow/addons/issues/2388#issuecomment-778422760.

**Comment by tabakg** [comment]
Thanks! I'll reach out to @tomerk with the internal thread I mentioned.

**Comment by tomerk** [comment]
Ecosystem review notes: Our internal discussion concluded addons would be a better home for this than tf core. It also does not conflict with plans for core tf.

**Comment by bhack** [comment]
@tomerk Do you have also an interested reviewer for this?

**Comment by tomerk** [comment]
Not off the top of my head but I imagine @tabakg can find one of the internal reviewers to do so.

**Comment by tabakg** [comment]
Thank you! My suggestion would be @penpornk (Penporn Koanantakool) and @bloops (Anudhyan Boral). I'll email them to make sure I found their correct Github accounts, and to let them know.

**Comment by bhack** [comment]
@tabakg Let us know if you have any feedback.

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
